cwltool/executors.py

```diff
@@ -87,22 +99,82 @@ class SingleJobExecutor(JobExecutor):
     def run_jobs(self,
                  t,  # type: Process
                  job_order_object,  # type: Dict[Text, Any]
+                 document,
+                 engineUUID,
+                 WorkflowRunID,
                  logger,
                  **kwargs  # type: Any
                  ):
+        reference_locations={} # type: Dict[Text, Any]
+        ProcessProvActivity=''
         jobiter = t.job(job_order_object,
                         self.output_callback,
                         **kwargs)
-
         try:
+            ro = kwargs.get("ro")
             for r in jobiter:
                 if r:
                     builder = kwargs.get("builder", None)  # type: Builder
+
                     if builder is not None:
                         r.builder = builder
                     if r.outdir:
                         self.output_dirs.add(r.outdir)
-                    r.run(**kwargs)
+                    if ro:
```

Comment:
I would suggest creating a separate subclass of `JobExecutor` here to keep the API simpler and cleaner since the number of changes and `if` clauses for a single feature is large enough.